# Preparación del entorno

In [1]:
!pip install wget

  Preparing metadata (setup.py) ... done
  Created wheel for wget: filename=wget-3.2-py3-none-any.whl size=9655 sha256=df767822ea4da3218c44d1d7c5ff3573ed192b444c61058faa58c5b4c0b0ccd9
  Stored in directory: /root/.cache/pip/wheels/8b/f1/7f/5c94f0a7a505ca1c81cd1d9208ae2064675d97582078e6c769
Successfully built wget


In [2]:
%%bash
pip install --upgrade duckdb
pip install --upgrade jupysql
pip install --upgrade duckdb-engine

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 16.5/16.5 MB 64.2 MB/s eta 0:00:00
  Attempting uninstall: duckdb
    Found existing installation: duckdb 0.8.1
    Uninstalling duckdb-0.8.1:
      Successfully uninstalled duckdb-0.8.1
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 87.6/87.6 kB 4.2 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 312.6/312.6 kB 14.5 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 376.5/376.5 kB 21.0 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.6/1.6 MB 37.8 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 43.1/43.1 kB 2.1 MB/s eta 0:00:00


In [3]:
import wget

url = "https://storage.googleapis.com/open-courses/dmeyf2023-8a1e/competencia_02_crudo.csv.gz"
ruta_destino = "/content/competencia_02_crudo.csv.gz"

wget.download(url, ruta_destino)

print("Archivo descargado en:", ruta_destino)

Archivo descargado en: /content/competencia_02_crudo.csv.gz


# Prueba con SQL

In [4]:
import duckdb
import pandas as pd

%load_ext sql
%config SqlMagic.autopandas = True
%config SqlMagic.feedback = False
%config SqlMagic.displaycon = False

%sql duckdb:////content/dmeyf.db

In [5]:
%%sql
create or replace table competencia_02 as
select
    *
from read_csv_auto('/content/competencia_02_crudo.csv.gz')

FloatProgress(value=0.0, layout=Layout(width='auto'), style=ProgressStyle(bar_color='black'))

Empty DataFrame
Columns: [Success]
Index: []

# Construcción de Clase_ternaria

In [6]:
%sql select * from competencia_02 limit 5

numero_de_cliente  foto_mes  active_quarter  cliente_vip  internet  \
0           29183981    201901               1            0         1   
1           29184630    201901               1            0         1   
2           29185587    201901               1            0         1   
3           29185646    201901               1            0         1   
4           29187499    201901               1            0         1   

   cliente_edad  cliente_antiguedad  mrentabilidad  mrentabilidad_annual  \
0            48                 171        8304.31              42993.54   
1            57                 296       18318.76              42597.26   
2            77                 296       20862.84              85481.31   
3            58                 231       25218.94             165317.78   
4            69                 349        5307.54              32374.08   

   mcomisiones  ...  Visa_madelantodolares  Visa_fultimo_cierre  Visa_mpagado  \
0      2501.04  ...                    0.0                    8          0.00   
1     14986.76  ...                    0.0                    1          0.00   
2      4228.40  ...                    0.0                    1          0.00   
3      4474.25  ...                    0.0                   22      69653.59   
4      4706.22  ...                    0.0                    8          0.00   

   Visa_mpagospesos  Visa_mpagosdolares  Visa_fechaalta  Visa_mconsumototal  \
0         -35773.32                2.33            5148            36602.13   
1         -67414.82             1149.57            3299            28503.06   
2              0.00                0.00            6790             5170.25   
3         -59422.00                0.00            7037            67991.90   
4          -4802.21                0.00            7596             4851.93   

   Visa_cconsumos  Visa_cadelantosefectivo  Visa_mpagominimo  
0              54                        0           5008.71  
1              10                        0           1395.87  
2               2                        0              0.00  
3              31                        0           2897.31  
4               3                        0            492.66  

[5 rows x 154 columns]

In [7]:
%%sql
select
    foto_mes
    , count(*) as cantidad -- cuenta cuantos casos hay en cada foto_mes
                           -- y lo guarda en un campo llamado cantidad
from competencia_02
group by foto_mes

foto_mes  cantidad
0     201901    124864
1     201902    125898
2     201903    126536
3     201904    127059
4     201905    127683
5     201906    129181
6     201907    130551
7     201908    132615
8     201909    134267
9     201910    136515
10    201911    138438
11    201912    140468
12    202001    143762
13    202002    146851
14    202003    148989
15    202004    149484
16    202005    150874
17    202006    153386
18    202007    155348
19    202008    156667
20    202009    158031
21    202010    159477
22    202011    160493
23    202012    161133
24    202101    161711
25    202102    162364
26    202103    163324
27    202104    163637
28    202105    164102
29    202106    164420
30    202107    164682

In [8]:
%%sql
create or replace table targets as
with periodos as (
    select distinct foto_mes from competencia_02
), clientes as (
    select distinct numero_de_cliente from competencia_02
), todo as (
    select numero_de_cliente, foto_mes from clientes cross join periodos
), clase_ternaria as (
    select
        t.numero_de_cliente
        , t.foto_mes
        , if(c.numero_de_cliente is null, 0, 1) as mes_0
        , lead(mes_0, 1) over (partition by t.numero_de_cliente order by foto_mes) as mes_1
        , lead(mes_0, 2) over (partition by t.numero_de_cliente order by foto_mes) as mes_2
        , CASE
    WHEN (mes_0 = 1) AND (mes_1 = 1) AND (mes_2 = 1) THEN 'CONTINUA'
    WHEN (mes_0 = 1) AND (mes_1 = 0) AND (mes_2 = 1) THEN 'CONTINUA'
    WHEN (mes_0 = 1) AND (mes_1 = 0) AND (mes_2 = 0) THEN 'BAJA+1'
    WHEN (mes_0 = 1) AND (mes_1 = 1) AND (mes_2 = 0) THEN 'BAJA+2'
    ELSE NULL END AS clase_ternaria
    from todo t
    left join competencia_02 c using (numero_de_cliente, foto_mes)
) select
  foto_mes
  , numero_de_cliente
  , clase_ternaria
from clase_ternaria where mes_0 = 1

FloatProgress(value=0.0, layout=Layout(width='auto'), style=ProgressStyle(bar_color='black'))

Empty DataFrame
Columns: [Success]
Index: []

In [9]:
%%sql
alter table competencia_02 add column clase_ternaria VARCHAR(10)

Empty DataFrame
Columns: [Success]
Index: []

In [10]:
%%sql
update competencia_02
set clase_ternaria = targets.clase_ternaria
from targets
where competencia_02.numero_de_cliente = targets.numero_de_cliente and competencia_02.foto_mes = targets.foto_mes;

Empty DataFrame
Columns: [Success]
Index: []

In [11]:
%%sql
SELECT clase_ternaria, COUNT(*) AS count_clase
FROM targets
GROUP BY clase_ternaria;

clase_ternaria  count_clase
0           None       329102
1         BAJA+2        18987
2       CONTINUA      4195938
3         BAJA+1        18783

In [12]:
%%sql
select
    foto_mes
    , count(*) as cantidad -- cuenta cuantos casos hay en cada foto_mes
                           -- y lo guarda en un campo llamado cantidad
from competencia_02
group by foto_mes

foto_mes  cantidad
0     201901    124864
1     201902    125898
2     201903    126536
3     201904    127059
4     201905    127683
5     201906    129181
6     201907    130551
7     201908    132615
8     201909    134267
9     201910    136515
10    201911    138438
11    201912    140468
12    202001    143762
13    202002    146851
14    202003    148989
15    202004    149484
16    202005    150874
17    202006    153386
18    202007    155348
19    202008    156667
20    202009    158031
21    202010    159477
22    202011    160493
23    202012    161133
24    202101    161711
25    202102    162364
26    202103    163324
27    202104    163637
28    202105    164102
29    202106    164420
30    202107    164682

# Feauture Engineering

Vamos por unos ejemplos simples. Que pasa si quiero agregar un feature que muestre el valor del periodo anterior?

In [13]:
%%sql
select
  numero_de_cliente
  , foto_mes
  , ctrx_quarter
  , lag(ctrx_quarter, 1) over (partition by numero_de_cliente order by foto_mes) as lag_1_ctrx_quarter
from competencia_02
limit 10

numero_de_cliente  foto_mes  ctrx_quarter  lag_1_ctrx_quarter
0           29188883    201901            43                 NaN
1           29188883    201902            45                43.0
2           29188883    201903            39                45.0
3           29188883    201904            38                39.0
4           29188883    201905            39                38.0
5           29188883    201906            41                39.0
6           29188883    201907            34                41.0
7           29188883    201908            30                34.0
8           29188883    201909            25                30.0
9           29188883    201910            30                25.0

Podemos calcular el delta (diferencia) entre el valor pasado y el presente, para uno o varios meses

In [14]:
%%sql
select
  numero_de_cliente
  , foto_mes
  , ctrx_quarter
  , lag(ctrx_quarter, 1) over (partition by numero_de_cliente order by foto_mes) as lag_1_ctrx_quarter
  , ctrx_quarter - lag_1_ctrx_quarter as delta_1_ctrx_quarter
  , ctrx_quarter - lag(ctrx_quarter, 2) over (partition by numero_de_cliente order by foto_mes) as lag_2_ctrx_quarter
from competencia_02
limit 10


numero_de_cliente  foto_mes  ctrx_quarter  lag_1_ctrx_quarter  \
0           29188883    201901            43                 NaN   
1           29188883    201902            45                43.0   
2           29188883    201903            39                45.0   
3           29188883    201904            38                39.0   
4           29188883    201905            39                38.0   
5           29188883    201906            41                39.0   
6           29188883    201907            34                41.0   
7           29188883    201908            30                34.0   
8           29188883    201909            25                30.0   
9           29188883    201910            30                25.0   

   delta_1_ctrx_quarter  lag_2_ctrx_quarter  
0                   NaN                 NaN  
1                   2.0                 NaN  
2                  -6.0                -4.0  
3                  -1.0                -7.0  
4                   1.0                 0.0  
5                   2.0                 3.0  
6                  -7.0                -5.0  
7                  -4.0               -11.0  
8                  -5.0                -9.0  
9                   5.0                 0.0

Si necesitamos ya no solo traer un valor del pasado, sino una secuencia de valores, por ejemplo para calcular la media móvil en los últimos 6 meses, se puede hacer fácilmente

In [15]:
%%sql
select
  numero_de_cliente
  , foto_mes
  , ctrx_quarter
  , avg(ctrx_quarter) over (partition by numero_de_cliente
                            order by foto_mes
                            rows between 6 preceding and current row) as avg_6_ctrx_quarter
from competencia_02
limit 10

numero_de_cliente  foto_mes  ctrx_quarter  avg_6_ctrx_quarter
0           29188883    201901            43           43.000000
1           29188883    201902            45           44.000000
2           29188883    201903            39           42.333333
3           29188883    201904            38           41.250000
4           29188883    201905            39           40.800000
5           29188883    201906            41           40.833333
6           29188883    201907            34           39.857143
7           29188883    201908            30           38.000000
8           29188883    201909            25           35.142857
9           29188883    201910            30           33.857143

- Para el cliente 29188883 para la foto 201910, se puede chequear que la nueva feature se construye con la siguiente cuenta (30 + 25 + 30 + 34 + 41 + 39 + 38) / 7
- Si embargo puede resultar incómodo escribir constantemente el over partition sobre todo si se buscan aplicar muchas veces para distintas funciones. Para reducir el código se puede usar la siguiente sintaxis

In [16]:
%%sql
select
  numero_de_cliente
  , foto_mes
  , ctrx_quarter
  , avg(ctrx_quarter) over ventana_6 as ctrx_quarter_media_6
  , max(ctrx_quarter) over ventana_6 as ctrx_quarter_max_6
  , min(ctrx_quarter) over ventana_6 as ctrx_quarter_min_6
from competencia_02
window ventana_6 as (partition by numero_de_cliente order by foto_mes rows between 6 preceding and current row)
limit 10

numero_de_cliente  foto_mes  ctrx_quarter  ctrx_quarter_media_6  \
0           29188883    201901            43             43.000000   
1           29188883    201902            45             44.000000   
2           29188883    201903            39             42.333333   
3           29188883    201904            38             41.250000   
4           29188883    201905            39             40.800000   
5           29188883    201906            41             40.833333   
6           29188883    201907            34             39.857143   
7           29188883    201908            30             38.000000   
8           29188883    201909            25             35.142857   
9           29188883    201910            30             33.857143   

   ctrx_quarter_max_6  ctrx_quarter_min_6  
0                  43                  43  
1                  45                  43  
2                  45                  39  
3                  45                  38  
4                  45                  38  
5                  45                  38  
6                  45                  34  
7                  45                  30  
8                  41                  25  
9                  41                  25

Para saber más que funciones tenemos disponibles, recomiendo ver los siguientes links:
- https://duckdb.org/docs/archive/0.8.1/sql/window_functions
- https://duckdb.org/docs/archive/0.8.1/sql/aggregates

Un caso más, que ni me voy a molestar en explicar que significa...

In [17]:
%%sql
select
  numero_de_cliente
  , foto_mes
  , ctrx_quarter
  ,regr_slope(ctrx_quarter, cliente_antiguedad) over ventana_6 as ctrx_quarter_slope_6
from competencia_02
window ventana_6 as (partition by numero_de_cliente order by foto_mes rows between 6 preceding and current row)
limit 10


numero_de_cliente  foto_mes  ctrx_quarter  ctrx_quarter_slope_6
0           29188883    201901            43                   NaN
1           29188883    201902            45              2.000000
2           29188883    201903            39             -2.000000
3           29188883    201904            38             -2.100000
4           29188883    201905            39             -1.500000
5           29188883    201906            41             -0.828571
6           29188883    201907            34             -1.250000
7           29188883    201908            30             -1.857143
8           29188883    201909            25             -2.250000
9           29188883    201910            30             -2.250000

... Alguno dirá "tenemos que escribir todo esto a mano? Son muchas variables!". Bueno no, use los conocimientos de programación para que la computadora trabaje para usted. Si tenemos una lista de campos

In [18]:

campos = ['active_quarter', 'cliente_vip', 'internet', 'cliente_edad', 'cliente_antiguedad', 'mrentabilidad', 'mrentabilidad_annual', 'mcomisiones', 'mactivos_margen', 'mpasivos_margen', 'cproductos', 'tcuentas', 'ccuenta_corriente', 'mcuenta_corriente_adicional', 'mcuenta_corriente', 'ccaja_ahorro', 'mcaja_ahorro', 'mcaja_ahorro_adicional', 'mcaja_ahorro_dolares', 'cdescubierto_preacordado', 'mcuentas_saldo', 'ctarjeta_debito', 'ctarjeta_debito_transacciones', 'mautoservicio', 'ctarjeta_visa', 'ctarjeta_visa_transacciones', 'mtarjeta_visa_consumo', 'ctarjeta_master', 'ctarjeta_master_transacciones', 'mtarjeta_master_consumo', 'cprestamos_personales', 'mprestamos_personales', 'cprestamos_prendarios', 'mprestamos_prendarios', 'cprestamos_hipotecarios', 'mprestamos_hipotecarios', 'cplazo_fijo', 'mplazo_fijo_dolares', 'mplazo_fijo_pesos', 'cinversion1', 'minversion1_pesos', 'minversion1_dolares', 'cinversion2', 'minversion2', 'cseguro_vida', 'cseguro_auto', 'cseguro_vivienda', 'cseguro_accidentes_personales', 'ccaja_seguridad', 'cpayroll_trx', 'mpayroll', 'mpayroll2', 'cpayroll2_trx', 'ccuenta_debitos_automaticos', 'mcuenta_debitos_automaticos', 'ctarjeta_visa_debitos_automaticos', 'mttarjeta_visa_debitos_automaticos', 'ctarjeta_master_debitos_automaticos', 'mttarjeta_master_debitos_automaticos', 'cpagodeservicios', 'mpagodeservicios', 'cpagomiscuentas', 'mpagomiscuentas', 'ccajeros_propios_descuentos', 'mcajeros_propios_descuentos', 'ctarjeta_visa_descuentos', 'mtarjeta_visa_descuentos', 'ctarjeta_master_descuentos', 'mtarjeta_master_descuentos', 'ccomisiones_mantenimiento', 'mcomisiones_mantenimiento', 'ccomisiones_otras', 'mcomisiones_otras', 'cforex', 'cforex_buy', 'mforex_buy', 'cforex_sell', 'mforex_sell', 'ctransferencias_recibidas', 'mtransferencias_recibidas', 'ctransferencias_emitidas', 'mtransferencias_emitidas', 'cextraccion_autoservicio', 'mextraccion_autoservicio', 'ccheques_depositados', 'mcheques_depositados', 'ccheques_emitidos', 'mcheques_emitidos', 'ccheques_depositados_rechazados', 'mcheques_depositados_rechazados', 'ccheques_emitidos_rechazados', 'mcheques_emitidos_rechazados', 'tcallcenter', 'ccallcenter_transacciones', 'thomebanking', 'chomebanking_transacciones', 'ccajas_transacciones', 'ccajas_consultas', 'ccajas_depositos', 'ccajas_extracciones', 'ccajas_otras', 'catm_trx', 'matm', 'catm_trx_other', 'matm_other', 'ctrx_quarter', 'Master_delinquency', 'Master_status', 'Master_mfinanciacion_limite', 'Master_Fvencimiento', 'Master_Finiciomora', 'Master_msaldototal', 'Master_msaldopesos', 'Master_msaldodolares', 'Master_mconsumospesos', 'Master_mconsumosdolares', 'Master_mlimitecompra', 'Master_madelantopesos', 'Master_madelantodolares', 'Master_fultimo_cierre', 'Master_mpagado', 'Master_mpagospesos', 'Master_mpagosdolares', 'Master_fechaalta', 'Master_mconsumototal', 'Master_cconsumos', 'Master_cadelantosefectivo', 'Master_mpagominimo', 'Visa_delinquency', 'Visa_status', 'Visa_mfinanciacion_limite', 'Visa_Fvencimiento', 'Visa_Finiciomora', 'Visa_msaldototal', 'Visa_msaldopesos', 'Visa_msaldodolares', 'Visa_mconsumospesos', 'Visa_mconsumosdolares', 'Visa_mlimitecompra', 'Visa_madelantopesos', 'Visa_madelantodolares', 'Visa_fultimo_cierre', 'Visa_mpagado', 'Visa_mpagospesos', 'Visa_mpagosdolares', 'Visa_fechaalta', 'Visa_mconsumototal', 'Visa_cconsumos', 'Visa_cadelantosefectivo', 'Visa_mpagominimo']


#Podemos hacer un script muy sencillo que nos genere el texto que hay que poner en una query para generar esas variables
nuevos_features = ""
for campo in campos:
  nuevos_features += f"\n, regr_slope({campo}, cliente_antiguedad) over ventana_6 as ctrx_{campo}_slope_6"
print(nuevos_features)


, regr_slope(active_quarter, cliente_antiguedad) over ventana_6 as ctrx_active_quarter_slope_6
, regr_slope(cliente_vip, cliente_antiguedad) over ventana_6 as ctrx_cliente_vip_slope_6
, regr_slope(internet, cliente_antiguedad) over ventana_6 as ctrx_internet_slope_6
, regr_slope(cliente_edad, cliente_antiguedad) over ventana_6 as ctrx_cliente_edad_slope_6
, regr_slope(cliente_antiguedad, cliente_antiguedad) over ventana_6 as ctrx_cliente_antiguedad_slope_6
, regr_slope(mrentabilidad, cliente_antiguedad) over ventana_6 as ctrx_mrentabilidad_slope_6
, regr_slope(mrentabilidad_annual, cliente_antiguedad) over ventana_6 as ctrx_mrentabilidad_annual_slope_6
, regr_slope(mcomisiones, cliente_antiguedad) over ventana_6 as ctrx_mcomisiones_slope_6
, regr_slope(mactivos_margen, cliente_antiguedad) over ventana_6 as ctrx_mactivos_margen_slope_6
, regr_slope(mpasivos_margen, cliente_antiguedad) over ventana_6 as ctrx_mpasivos_margen_slope_6
, regr_slope(cproductos, cliente_antiguedad) over venta

In [ ]:
%%sql
copy competencia_02 to '/content/competencia_02.csv.gz' (FORMAT CSV, HEADER)

FloatProgress(value=0.0, layout=Layout(width='auto'), style=ProgressStyle(bar_color='black'))

Empty DataFrame
Columns: [Success]
Index: []